### ISSUE TITLE
Getting `ERR_PNPM_HARDLINK_FAILED Error: EEXIST: file already exists, link` when pnpm install

### ISSUE THREAD

**Issue by llamington** [pnpm/pnpm#7713]
### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### Which area(s) of pnpm are affected? (leave empty if unsure)

CLI

### Reproduction steps

* Set up a project with a pnpm package manager configuration.
* Include a git dependency in your package.json that references a pnpm monorepo.
* Run pnpm install in a build pipeline environment (e.g., CodeMagic MacOS M1).
* Observe that approximately 50% of the time, you encounter the ERR_PNPM_HARDLINK_FAILED error with the specific message EEXIST: file already exists, link.

### Describe the Bug

When running pnpm install in a CodeMagic build pipeline environment with Node 21, there is a recurring issue where approximately 50% of the time, the installation process fails with the error ERR_PNPM_HARDLINK_FAILED. This error occurs specifically when attempting to create a hard link for a file that already exists. The error message points to a specific file path within the node_modules directory, which seems to be related to the installation of a git dependency referencing a monorepo.
Here's the entire message:
```
Error: EEXIST: file already exists, link '/Users/builder/Library/pnpm/store/v3/tmp/_tmp_6353_fd2fc06335afb29e32c6e23f10f514aa/node_modules/@google-cloud/datastore/node_modules/google-gax/node_modules/@grpc/proto-loader/node_modules/protobufjs/dist/minimal/protobuf.js' -> 'functions/node_modules/google-gax/node_modules/@grpc/proto-loader/node_modules/protobufjs/dist/minimal/protobuf.js'
```

### Expected Behavior

`pnpm install` completes without an error code

### Which Node.js version are you using?

21

### Which operating systems have you used?

- [X] macOS
- [ ] Windows
- [ ] Linux

**Comment by Yidoon** [comment]
I encountered a similar error during the Jenkins build process when executing the `pnpm i` command. 

 The error message appears as follows:
ERR_PNPM_LINKING_FAILED  Error: EEXIST: file already exists, rename '/var/lib/jenkins/workspace/projectName/node_modules/@lingui/cli/node_modules/esbuild/node_modules' -> '/var/lib/jenkins/workspace/projectName/node_modules/@lingui/cli/node_modules/esbuild_tmp_3229/node_modules'

- node version: v18.0.0
- pnpm version: 8.10.5
- monorepo: pnpm workspace
- linux

 I addressed the issue using several commands before pnpm i
rm -rf node_modules 
rm -rf ./web/node_modules
rm -rf pnpm-lock.yaml

 However, I understand that this isn't the optimal solution. 
: )

**Comment by literalpie** [comment]
I'm also seeing this:
```
 ERR_PNPM_EEXIST  EEXIST: file already exists, link '/__w/ufe/ufe/node_modules/@storybook/core-common/node_modules/esbuild/LICENSE.md' -> 'node_modules/@yarnpkg/esbuild-plugin-pnp/node_modules/esbuild/LICENSE.md'
```

I wonder if it could have something to do with having multiple versions of esbuild in the dependency tree? The two instances of esbuild mentioned in the error are the same version, but there's a different version of esbuild elsewhere in my deps.

**Comment by literalpie** [comment]
I'm able to work around this by using the pnpm cache in setup-node:
```
- name: Setup Node
      uses: actions/setup-node@v3
      with:
        cache: 'pnpm'
```

This isn't a great workaround because it adds 5 minutes to the "Post Job Cleanup" step on the first run of each branch, but it at least unblocks me.

**Comment by literalpie** [comment]
I had hoped #8712 would fix this, but it did not

**Comment by onurusluca** [comment]
Got this too on a Nuxt frontend.

```
ERR_PNPM_EEXIST  EEXIST: file already exists, rename '/vercel/path1/node_modules/@parcel/watcher-wasm/node_modules' -> '/vercel/path1/node_modules/@parcel/watcher-wasm_tmp_115/node_modules'
Error: Command "pnpm install" exited with 1
```

**Comment by zkochan** [comment]
Closing for lack of a reproduction on a supported version.

The store and the linker are native code now. Please open a new issue with the full output if a hard link still fails intermittently.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
